### PR TITLE
Better dummies

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR reworks how the dummy objects work by using a metaclass to build them, and make sure that any classmethod used will raise the appropriate error message.

Fixes the bad error message when trying to do something like:
```python
FlaxVisionEncoderDecoderModel.from_encoder_decoder_pretrained("vit", "gpt2")
```
with Flax not installed.

### CODE REVIEW

File: utils/check_dummies.py [hunk 1]
```diff
@@ -176,7 +112,8 @@ def create_dummy_files():
     for backend, objects in backend_specific_objects.items():
         backend_name = "[" + ", ".join(f'"{b}"' for b in backend.split("_and_")) + "]"
         dummy_file = "# This file is autogenerated by the command `make fix-copies`, do not edit.\n"
-        dummy_file += "from ..file_utils import requires_backends\n\n"
+        dummy_file += "# flake8: noqa\n"
```
Comment (sgugger) [inline utils/check_dummies.py:115]: This line is necessary since `DummyObject` is sometimes not used in those files and then flake8 is unhappy.

**Review by LysandreJik — APPROVED** [review]
Nice, thank you @sgugger!

Eventually it would be nice to write a guide regarding how to contribute to this/solve common issues, as it is quite complex to understand how everything works/fits together.

**Review by LysandreJik — APPROVED** [review]
Hadn't understood the PR in its entirety, thanks for showing me around.

I understand that one of the limitations of such an approach is that doing a typo in an attribute/method yields the error anyway:
```
>>> CLIPFeatureExtractor.issue
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/lysandre/Workspaces/Python/transformers/src/transformers/file_utils.py", line 838, in __getattr__
    requires_backends(cls, cls._backends)
  File "/home/lysandre/Workspaces/Python/transformers/src/transformers/file_utils.py", line 831, in requires_backends
    raise ImportError("".join([BACKENDS_MAPPING[backend][1].format(name) for backend in backends]))
ImportError:
CLIPFeatureExtractor requires the PIL library but it was not found in your environment. You can install it with pip:
`pip install pillow`

>>>
```

I don't think that's much of an issue though as the error message is still relevant.

LGTM

**Comment by sgugger** [comment]
Yes, it will send the imprt error for any attribute, but at the same time the object should just not be used if the right frameworks are not installed.

**Comment by sgugger** [comment]
Failure is non-related so merging.

**Pull request merged**